graph-builder/src/config/unified.rs

```diff
@@ -0,0 +1,156 @@
+//! Unified configuration.
+
+use super::cli::CliOptions;
+use super::file::FileOptions;
+use failure::Fallible;
+use std::collections::{HashMap, HashSet};
+
+static DEFAULT_VERBOSITY: u8 = 0;
+
+static DEFAULT_SERV_ADDR: &str = "127.0.0.1";
+static DEFAULT_SERV_PORT: u16 = 8080;
+static DEFAULT_SERV_PREFIX: &str = "";
+
+static DEFAULT_REGISTRY_PERIOD: u64 = 30;
+static DEFAULT_REGISTRY_CREDENTIALS_PATH: &str = "";
+static DEFAULT_REGISTRY_MANIFESTREF_KEY: &str = "com.openshift.upgrades.graph.release.manifestref";
+static DEFAULT_REGISTRY_URL: &str = "http://localhost:5000";
+static DEFAULT_REGISTRY_REPO: &str = "openshift";
+
+static DEFAULT_STATUS_ADDR: &str = "127.0.0.1";
+static DEFAULT_STATUS_PORT: u16 = 9080;
+
+macro_rules! maybe_assign {
+    ( $dst:expr, $src:expr ) => {{
+        if let Some(x) = $src {
+            $dst = x;
+        };
+    }};
+}
+
+/// Top-level configuration (before semantic validation).
+#[derive(Debug)]
+pub struct UnifiedConfig {
```

Comment:
Is there a reason for why we can't have `Option`s in here?

Reply:
Is there anything specific that you think should be an `Option`?

This struct holds configuration entries, after merging together defaults+toml+cli. Unless something here requires a real `None`, most of the entries will have real values (either from user-options or from defaults).

Reply:
I would imagine that this struct could be a mere composition of the ones we already have for CLI and file, so this would resemble the same structure everywhere. I understand this places the burden of handling the `Option`s again in the code :thinking: